samples: nrf5340: Add sample showing how to build multicore app (#7295)
nrfconnect/sdk-nrf: samples/nrf5340/multicore/README.rst

.. _nrf5340_multicore:

nRF5340: Multicore application
###############################

.. contents::
   :local:
   :depth: 2

> You can use this sample to see how to build an application on the network core as a child image from the source.

Reviewer: ```suggestion
You can use this sample to see how to build an application on the network core as a child image from the sample source files.
```

Requirements
************

The sample supports the following development kits:

.. table-from-rows:: /includes/sample_board_rows.txt
   :header: heading
   :rows: nrf5340dk_nrf5340_cpuapp

Overview
********

The sample demonstrates how to build a multicore application with the :c:macro:`add_child_image` CMake macro.
The configuration of the application follows the process outlined in :ref:`ug_multi_image_defining` on the :ref:`ug_multi_image` page.

> To meet the configuration requirements, the sample comes with the following additional source files:

Reviewer: ```suggestion
The sample comes with the following additional files:
```

> This sample shows how to inform the build system about dedicated sources for the child image (in the :file:`zephyr` and :file:`aci` directories).

Reviewer 2: ```suggestion
If you use the :c:macro:`add_child_image` macro, this sample informs the build system about dedicated sources for the child image (in the :file:`zephyr` and :file:`aci` directories).
```
Added this, as the issues is only when the user uses `add_child_image` macro. Hope this didn't change the meaning.

Author: It cant be like that. I will keep the old version.

   * :file:`aci/CMakeLists.txt` - This file is used to add the child image and contains the :c:macro:`add_child_image` macro.
   * :file:`zephyr/module.yml` - This file contains information about the location of the :file:`CMakeLists.txt` file that invokes :c:macro:`add_child_image` that is required for the build system.

>    * :file:`Kconfig` - Custom file that allows to add child image only for application core.

Author: Should be fine now, please double check.

     Additionally, it enables options required by multi-image build.

Both the application and network cores use the same :file:`main.c` that prints the name of the DK on which the application is programmed.

> For more information about multi-image builds, see :ref:`ug_multi_image`.

Reviewer: After discussing on Teams, let's remove line 25.
We can rephrase this one like follows:
```suggestion
For general information about multi-image builds, see :ref:`ug_multi_image`.
```

.. _nrf5340_multicore:

nRF5340: Multi core application
###############################

.. contents::
   :local:
   :depth: 2

> You can use this sample to see how to build an application with the network core build as child image from soruce.

Reviewer 3: Just a thought.  Shouldn't we say that we can use this as a sample for running application on any Nordic SoC with 2 cores. It will be ready for Haltium SoCs.
  "... can use this sample to see how to build application on two cores. In this example we use..."

Requirements
************

The sample supports the following development kits:

.. table-from-rows:: /includes/sample_board_rows.txt
   :header: heading
   :rows: nrf5340dk_nrf5340_cpuapp

Overview
********

The sample application shows how to build multi core application, with use of ``add_child_image`` cmake macro.
Note the application source dir structure and additional files:

   * ``aci/CMakeLists.txt`` - adding child image.
   * ``zephyr/module.yml`` - information about location of ``CMakeLists.txt`` file that invokes ``add_child_image``.
     Required for build system.

>    * Custom ``Kconfig`` file that allows to add childe image only for application core.

Reviewer 3: childe/child

> The sample application shows how to build multi core application, with use of ``add_child_image`` cmake macro.
> Note the application source dir structure and additional files:
> 
>    * ``aci/CMakeLists.txt`` - adding child image.
>    * ``zephyr/module.yml`` - information about location of ``CMakeLists.txt`` file that invokes ``add_child_image``.
>      Required for build system.
>    * Custom ``Kconfig`` file that allows to add child image only for application core.
>      Additionally it enables options required by multi image build.

Reviewer: ```suggestion
The sample demonstrates how to build a multicore application with the :c:macro:`add_child_image` CMake macro.

The configuration of the application follows the process outlined in :ref:`ug_multi_image_defining` on the :ref:`ug_multi_image` page.
To meet the configuration requirements, the sample comes with the following additional source files:

   *  :file:`aci/CMakeLists.txt` - This file is used to add the child image and contains the :c:macro:`add_child_image` macro.
   * :file:`zephyr/module.yml` - This file contains information about the location of the :file:`CMakeLists.txt` file that invokes :c:macro:`add_child_image` that is required for the build system.
   * :file:`Kconfig` - Custom file that allows to add child image only for application core.
     Additionally, it enables options required by multi-image build.
```
What about the multi-image variables described in the Multi-image guide?

>      Additionally it enables options required by multi image build.

Author: What about those variables? They work as usual. The child_image is named `cpunet` as could be seen in `aci/CMakeLists.txt` file when using `add_child_image` macro.

> Both cores uses the same ``main.c``, which prints name of the board its run on.

Reviewer: ```suggestion
Both the application and network cores use the same :file:`main.c` that prints the name of the DK on which the application is programmed.
```

For more information about multi image builds, see :ref:`ug_multi_image`.

> Building and running

Reviewer: In the multi-image user guide, we describe changes in the build process related to using multi-image. How does using multi-image in this sample affect the building process?

Author: That's a bit complicated, i will try to be brief.

Once you call `find_package(Zephyr REQUIRED HINTS $ENV{ZEPHYR_BASE})` in Applications `CMakeLists.txt`, NCS repository is scanned for related dependencies, child_images in particular. Child images are selected (and build) depending on selected options in root image (for our case its application image). Apart from child image mechanism, NCS has many others, e.g. Partition manager.

In this build system one thing was not considered - if the user wants to use `add_child_image` by himself (from applications (`CMakeLists.txt`). Because of how build system is build without those "magic files" i've added (`aci/CMakeLists.txt` and `zephyr/module.yml`), the build invokes some cmake files in wrong order (partition manager). What results in build error.

What this sample (and "magic files") does, it tricks the build system so that it call `partition_manager.cmake` alongside other files (`multi_image.cmake`) in correct order. That allows to build the application correctly.

I hope that clears things a bit.

Reviewer: Thanks! Let's add this explanation to line 36, after the new line 35.

```
When building any application for the nRF5340 SoC, the build system adds the child images based on the options selected for the parent image.
This sample overrides this process by specifying its dedicated sources for the child images.
To do this, it points the build system to the correct child image location in the :file:`aci/CMakeLists.txt` file by using :file:`zephyr/module.yml`.
```

********************

> .. |sample path| replace:: :file:`samples/ipc/ipc_service`

Author: Sorry, i forgot to update it, as i was creating those file. Fill fix it. Thanks for noticing!

.. include:: /includes/build_and_run.txt

Testing
=======

To test the sample, complete the following steps:

1. Build and program the sample.
#. Reset the kit.

> #. Observe the console output of both cores, it should be similar to the followingon UART for application core::

Reviewer 3: followingon/following on

      *** Booting Zephyr OS build v2.7.99-ncs1-2193-gd359a86abf14  ***
      Hello world from nrf5340dk_nrf5340_cpuapp

> #. Observe the console output of both cores, it should be similar to the followingon UART for network core::

Reviewer 3: followingon/following on

      *** Booting Zephyr OS build v2.7.99-ncs1-2193-gd359a86abf14  ***
      Hello world from nrf5340dk_nrf5340_cpunet

Dependencies
************

> In addition, it uses the following Zephyr subsystems:

Author: That was copy paste. There were section about NCS modules. I will just remove `in addition`.

> In addition, it uses the following Zephyr subsystems:

Author: Actually there are no dependencies. Logging subsystem is not used too. I will remove this section.
